lib/ImportFromC/HeaderToModel.cpp

```diff
@@ -235,137 +225,105 @@ DeclVisitor::getOrCreatePrimitive(const BuiltinType *UnderlyingBuiltin,
+    return model::PrimitiveType::makeUnsigned(1);
 
-  return std::nullopt;
-}
-
-namespace TDKind = model::TypeDefinitionKind;
-std::optional<model::DefinitionReference>
-DeclVisitor::getTypeByNameOrID(llvm::StringRef Name,
-                               TypeDefinitionKind::Values Kind) {
-  const bool IsStruct = Kind == TDKind::StructDefinition;
-  const bool IsUnion = Kind == TDKind::UnionDefinition;
-  const bool IsEnum = Kind == TDKind::EnumDefinition;
-  const bool IsCABI = Kind == TDKind::CABIFunctionDefinition;
-  const bool IsRaw = Kind == TDKind::RawFunctionDefinition;
+  case BuiltinType::Char_U:
+  case BuiltinType::UChar:
+    return model::PrimitiveType::makeSigned(1);
 
-  // Find by name first.
-  for (auto &Type : Model->TypeDefinitions()) {
-    if (IsStruct and not llvm::isa<model::StructDefinition>(Type.get()))
-      continue;
+  case BuiltinType::Void:
+    return model::PrimitiveType::makeVoid();
 
-    if (IsUnion and not llvm::isa<model::UnionDefinition>(Type.get()))
-      continue;
+  case BuiltinType::Float16:
+    return model::PrimitiveType::makeFloat(2);
 
-    if (IsEnum and not llvm::isa<model::EnumDefinition>(Type.get()))
-      continue;
+  case BuiltinType::Float:
+    return model::PrimitiveType::makeFloat(4);
 
-    if (IsCABI and not llvm::isa<model::CABIFunctionDefinition>(Type.get()))
-      continue;
+  case BuiltinType::Double:
+    return model::PrimitiveType::makeFloat(8);
 
-    if (IsRaw and not llvm::isa<model::RawFunctionDefinition>(Type.get()))
-      continue;
+  case BuiltinType::Float128:
+  case BuiltinType::LongDouble:
+    return model::PrimitiveType::makeFloat(16);
 
-    if (Type->CustomName() == Name)
-      return Model->getDefinitionReference(Type.get());
+  default:
+    revng_log(Log, "Unable to handle a primitive type");
   }
 
+  return nullptr;
+}
+
+template<std::derived_from<model::TypeDefinition> T>
+  requires(!std::is_same_v<T, model::TypeDefinition>)
+model::UpcastableType DeclVisitor::makeTypeByNameOrID(llvm::StringRef Name) {
+  // Try to find by name first.
+  for (auto &Type : Model->TypeDefinitions())
+    if (llvm::isa<T>(Type.get()))
+      if (Type->CustomName() == Name)
+        return Model->makeType(Type->key());
+
+  // Getting here means we didn't manage to find it,
+  // let's try parsing the name.
   size_t LocationOfID = Name.rfind("_");
 
   if (LocationOfID != std::string::npos) {
     std::string ID = std::string(Name.substr(LocationOfID + 1));
-    uint64_t TypeID;
-
-    std::istringstream TheStream(ID);
-    TheStream >> TypeID;
-
-    auto KeyType = model::TypeDefinition::Key{ TypeID, Kind };
-
-    auto TheType = Model->getDefinitionReference(KeyType);
-    if (TheType.get())
-      return TheType;
+    llvm::Expected<uint64_t> DeserializedID = deserialize<uint64_t>(ID);
+    if (DeserializedID)
+      return Model->makeType(model::TypeDefinition::Key{ *DeserializedID,
+                                                         T::AssociatedKind });
```

Comment:
I think you're not handling the error case.
You need to explicitly state "I don't care about the error". Basically, in case of failure, I think this code will abort because you didn't consume the error.

You have to use `llvm::consumeError` or `logAllUnhandledErrors` to a logger, but you need to check if that consume the errors too or not.

We have probably many places in our codebase where we get this wrong.

Reply:
Fixed (only here, didn't look for other offenders).